drivers/sensor/bmp581/bmp581.c

```diff
@@ -0,0 +1,561 @@
+/*
+ * Copyright (c) 2022 Badgerd Technologies B.V.
+ * Copyright (c) 2023 Metratec GmbH
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "bmp581.h"
+
+#include <math.h>
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/init.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/check.h>
+
+LOG_MODULE_REGISTER(bmp581, CONFIG_SENSOR_LOG_LEVEL);
+
+#if DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) == 0
+#warning "BMP581 driver enabled without any devices"
+#endif
+
+static int power_up_check(const struct device *dev);
+static int get_nvm_status(uint8_t *nvm_status, const struct device *dev);
+static int get_interrupt_status(uint8_t *int_status, const struct device *dev);
+static int validate_chip_id(struct bmp581_data *drv);
+static int get_osr_odr_press_config(struct bmp581_osr_odr_press_config *osr_odr_press_cfg,
+				    const struct device *dev);
+static int set_osr_config(const struct sensor_value *osr, enum sensor_channel chan,
+			  const struct device *dev);
+static int set_odr_config(const struct sensor_value *odr, const struct device *dev);
+static int soft_reset(const struct device *dev);
+static int set_iir_config(const struct sensor_value *iir, const struct device *dev);
+static int get_power_mode(enum bmp5_powermode *powermode, const struct device *dev);
+static int set_power_mode(enum bmp5_powermode powermode, const struct device *dev);
+
+static int set_power_mode(enum bmp5_powermode powermode, const struct device *dev)
+{
+	struct bmp581_config *conf = (struct bmp581_config *)dev->config;
+	int ret = BMP5_OK;
+	uint8_t odr = 0;
+	enum bmp5_powermode current_powermode;
+
+	CHECKIF(dev == NULL) {
+		return -EINVAL;
+	}
+
+	ret = get_power_mode(&current_powermode, dev);
+	if (ret != BMP5_OK) {
+		LOG_ERR("Couldnt set the power mode because something went wrong when getting the "
+			"current power mode.");
+		return ret;
+	}
+
+	if (current_powermode != BMP5_POWERMODE_STANDBY) {
+		/*
+		 * Device should be set to standby before transitioning to forced mode or normal
+		 * mode or continuous mode.
+		 */
+
+		ret = i2c_burst_read_dt(&conf->i2c, BMP5_REG_ODR_CONFIG, &odr, 1);
```

Comment:
For single byte reads and writes, you can use `i2c_reg_read_byte_dt` and `i2c_reg_write_byte` respectively